File: 2512.py
import sys
sys.stdin = open('input.txt')
n = int(input())
list = list(map(int, input().split()))
m = int(input())
r = max(list)
l = 0
res = 0
while l <= r:
    mid = (r+l)//2
    sum = 0
    for i in range(n):
        if mid >= list[i]:
            sum += list[i]
        else:
            sum += mid
    if sum <= m:
        l = mid + 1
        res = mid
    else:
        r = mid - 1
print(res)

File: beakJoon/2512.js
const inputs = require("fs").readFileSync("input.txt").toString().split("\n");
let n = parseInt(inputs[0]);
let budgets = inputs[1].split(" ").map((e) => parseInt(e));
let M = parseInt(inputs[2]);

function solution(budgets, M) {
  const max = (_) => _.reduce((_, $) => (_ > $ ? _ : $));
  var answer = 0;
  let l = 0;
  let r = max(budgets);
  let mid = null;
  let sum = 0;
  while (l <= r) {
    mid = Math.floor((l + r) / 2);
    sum = 0;
    for (let i = 0; i < budgets.length; i++) {
      if (mid >= budgets[i]) sum += budgets[i];
      else sum += mid;
    }
    if (sum <= M) {
      l = mid + 1;
      answer = mid;
    } else {
      r = mid - 1;
    }
  }
  return answer;
}

console.log(solution(budgets, M));
